# Introduction

Hopefully, this short tutorial can show you a lot of different commands that will help you gain the most insights into your dataset. 

In [2]:
import pandas as pd
from src.utils import load_data_from_google_drive

# Loading in Data

The first step in any ML problem is identifying what format your data is in, and then loading it into whatever framework you're using. For Kaggle compeitions, a lot of data can be found in CSV files, so that's the example we're going to use. 

We're going to be looking at a sports dataset that shows the results from NCAA basketball games from 1985 to 2016. This dataset is in a CSV file, and the function we're going to use to read in the file is called **pd.read_csv()**. This function returns a **dataframe** variable. The dataframe is the golden jewel data structure for Pandas. It is defined as "a two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns)".

Just think of it as a table for now. 

In [3]:
df = load_data_from_google_drive(url='https://drive.google.com/file/d/184JcLbSpArA_uq0DgAv2k892KChJVPHt/view?usp=share_link')

In [4]:
df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
1         1985      25   1106      77   1354      70    H      0
2         1985      25   1112      63   1223      56    H      0
3         1985      25   1165      70   1432      54    H      0
4         1985      25   1192      86   1447      74    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

# The Basics

Now that we have our dataframe in our variable df, let's look at what it contains. We can use the function **head()** to see the first couple rows of the dataframe (or the function **tail()** to see the last few rows).

In [5]:
df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0
4    1985      25   1192      86   1447      74    H      0

In [6]:
df.tail()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

We can see the dimensions of the dataframe using the the **shape** attribute

In [7]:
df.shape

(145289, 8)

We can also extract all the column names as a list, by using the **columns** attribute and can extract the rows with the **index** attribute

In [8]:
df.columns.tolist()

['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot']

In order to get a better idea of the type of data that we are dealing with, we can call the **describe()** function to see statistics like mean, min, etc about each column of the dataset. 

In [9]:
df.describe()

Season         Daynum          Wteam         Wscore  \
count  145289.000000  145289.000000  145289.000000  145289.000000   
mean     2001.574834      75.223816    1286.720646      76.600321   
std         9.233342      33.287418     104.570275      12.173033   
min      1985.000000       0.000000    1101.000000      34.000000   
25%      1994.000000      47.000000    1198.000000      68.000000   
50%      2002.000000      78.000000    1284.000000      76.000000   
75%      2010.000000     103.000000    1379.000000      84.000000   
max      2016.000000     132.000000    1464.000000     186.000000   

               Lteam         Lscore          Numot  
count  145289.000000  145289.000000  145289.000000  
mean     1282.864064      64.497009       0.044387  
std       104.829234      11.380625       0.247819  
min      1101.000000      20.000000       0.000000  
25%      1191.000000      57.000000       0.000000  
50%      1280.000000      64.000000       0.000000  
75%      1375.000000      72.000000       0.000000  
max      1464.000000     150.000000       6.000000

Okay, so now let's looking at information that we want to extract from the dataframe. Let's say I wanted to know the max value of a certain column. The function **max()** will show you the maximum values of all columns

In [10]:
df.max()

Season    2016
Daynum     132
Wteam     1464
Wscore     186
Lteam     1464
Lscore     150
Wloc         N
Numot        6
dtype: object

Then, if you'd like to specifically get the max value for a particular column, you pass in the name of the column using the bracket indexing operator

In [11]:
df['Wscore'].max()

186

If you'd like to find the mean of the Losing teams' score. 

In [12]:
df['Lscore'].mean()

64.49700940883343

But what if that's not enough? Let's say we want to actually see the game(row) where this max score happened. We can call the **argmax()** function to identify the row index

In [13]:
df['Wscore'].argmax()

24970

One of the most useful functions that you can call on certain columns in a dataframe is the **value_counts()** function. It shows how many times each item appears in the column. This particular command shows the number of games in each season

In [14]:
df['Season'].value_counts()

2016    5369
2014    5362
2015    5354
2013    5320
2010    5263
2012    5253
2009    5249
2011    5246
2008    5163
2007    5043
2006    4757
2005    4675
2003    4616
2004    4571
2002    4555
2000    4519
2001    4467
1999    4222
1998    4167
1997    4155
1992    4127
1991    4123
1996    4122
1995    4077
1994    4060
1990    4045
1989    4037
1993    3982
1988    3955
1987    3915
1986    3783
1985    3737
Name: Season, dtype: int64

**Q**: How many unique seasons are there in the dataset? Use the nunique() function.

In [15]:
df['Season'].nunique()

32

**Q**: Find the team with the most wins. Use the value_counts() function on the Wteam column.

In [16]:
df['Wteam'].value_counts()

1181    819
1242    804
1246    765
1314    761
1112    746
       ... 
1101     18
1446     14
1118      6
1289      6
1327      3
Name: Wteam, Length: 364, dtype: int64

# Acessing Values   

Then, in order to get attributes about the game, we need to use the **iloc[]** function. Iloc is definitely one of the more important functions. The main idea is that you want to use it whenever you have the integer index of a certain row that you want to access. As per Pandas documentation, iloc is an "integer-location based indexing for selection by position."

In [17]:
df.iloc[[df['Wscore'].argmax()]]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
24970    1991      68   1258     186   1109     140    H      0

Let's take this a step further. Let's say you want to know the game with the highest scoring winning team (this is what we just calculated), but you then want to know how many points the losing team scored. 

In [18]:
df.iloc[[df['Wscore'].argmax()]]['Lscore']

24970    140
Name: Lscore, dtype: int64

When you see data displayed in the above format, you're dealing with a Pandas **Series** object, not a dataframe object.

In [19]:
type(df.iloc[[df['Wscore'].argmax()]]['Lscore'])

pandas.core.series.Series

In [20]:
type(df.iloc[[df['Wscore'].argmax()]])

pandas.core.frame.DataFrame

The following is a summary of the 3 data structures in Pandas (Haven't ever really used Panels yet)

![](DataStructures.png)

When you want to access values in a Series, you'll want to just treat the Series like a Python dictionary, so you'd access the value according to its key (which is normally an integer index)

In [21]:
df.iloc[[df['Wscore'].argmax()]]['Lscore'][24970]

140

The other really important function in Pandas is the **loc** function. Contrary to iloc, which is an integer based indexing, loc is a "Purely label-location based indexer for selection by label". Since all the games are ordered from 0 to 145288, iloc and loc are going to be pretty interchangable in this type of dataset

In [22]:
df.iloc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

In [23]:
df.loc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0

Notice the slight difference in that iloc is exclusive of the second number, while loc is inclusive. 

Below is an example of how you can use loc to acheive the same task as we did previously with iloc

In [24]:
df.loc[df['Wscore'].argmax(), 'Lscore']

140

A faster version uses the **at()** function. At() is really useful wheneever you know the row label and the column label of the particular value that you want to get. 

In [25]:
df.at[df['Wscore'].argmax(), 'Lscore']

140

If you'd like to see more discussion on how loc and iloc are different, check out this great Stack Overflow post: http://stackoverflow.com/questions/31593201/pandas-iloc-vs-ix-vs-loc-explanation. Just remember that **iloc looks at position** and **loc looks at labels**. Loc becomes very important when your row labels aren't integers. 

# Sorting

Let's say that we want to sort the dataframe in increasing order for the scores of the losing team

In [26]:
df.sort_values('Lscore').head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
100027    2008      66   1203      49   1387      20    H      0
49310     1997      66   1157      61   1204      21    H      0
89021     2006      44   1284      41   1343      21    A      0
85042     2005      66   1131      73   1216      22    H      0
103660    2009      26   1326      59   1359      22    H      0

**Q**: Make three dataframes that are sorted by season, winning team, and winning score respectively. Then, Using iloc, select the rows from index 100 to 200 and the columns for season, winning team, and winning score, respectively. 

In [27]:

sorted_by_season = df.sort_values('Season')

sorted_by_winning_team = df.sort_values('Wteam')

sorted_by_winning_score = df.sort_values('Wscore')

selected_rows_columns = df.iloc[100:201, [0, 1, 2]]

**Q**: From these three subsets you obtained above, find the season and winning team for the game with the highest winning score.

In [28]:
highest_winning_score_game = sorted_by_winning_score.iloc[-1]  # Select last row
season = highest_winning_score_game['Season']
winning_team = highest_winning_score_game['Wteam']

print('Season: ', season)
print('Winning Team: ', winning_team)

Season:  1991
Winning Team:  1258


# Filtering Rows Conditionally

Now, let's say we want to find all of the rows that satisy a particular condition. For example, I want to find all of the games where the winning team scored more than 150 points. The idea behind this command is you want to access the column 'Wscore' of the dataframe df (df['Wscore']), find which entries are above 150 (df['Wscore'] > 150), and then returns only those specific rows in a dataframe format (df[df['Wscore'] > 150]).

In [29]:
df[df['Wscore'] > 150]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
5269     1986      75   1258     151   1109     107    H      0
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
16040    1989      40   1328     152   1331     122    H      0
16853    1989      68   1258     162   1109     144    A      0
17867    1989      92   1258     181   1109     150    H      0
19653    1990      30   1328     173   1109     101    H      0
19971    1990      38   1258     152   1109     137    A      0
20022    1990      40   1116     166   1109     101    H      0
22145    1990      97   1258     157   1362     115    H      0
23582    1991      26   1318     152   1258     123    N      0
24341    1991      47   1328     172   1258     112    H      0
24970    1991      68   1258     186   1109     140    H      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
40060    1995      32   1375     156   1341     114    H      0
52600    1998      33   1395     153   1410      87    H      0

This also works if you have multiple conditions. Let's say we want to find out when the winning team scores more than 150 points and when the losing team scores below 100. 

In [30]:
df[(df['Wscore'] > 150) & (df['Lscore'] < 100)]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
52600    1998      33   1395     153   1410      87    H      0

**Q**: Create a new column in the DataFrame called 'ScoreDifference' which is the absolute difference between the winning score and the losing score. Filter the DataFrame to only include games where the 'ScoreDifference' is greater than the average 'ScoreDifference' for all games.

In [31]:
# Step 1: Create the 'ScoreDifference' column
df['ScoreDifference'] = abs(df['Wscore'] - df['Lscore'])

# Step 2: Calculate the average 'ScoreDifference' for all games
average_diff = df['ScoreDifference'].mean()

# Step 3: Filter the DataFrame to include games with 'ScoreDifference' greater than the average
filtered_df = df[df['ScoreDifference'] > average_diff]


**Q**: From this filtered DataFrame, find the season and teams involved in the game with the highest 'ScoreDifference'.

In [32]:
# Find the row with the highest 'ScoreDifference'
max_diff_row = filtered_df[filtered_df['ScoreDifference'] == filtered_df['ScoreDifference'].max()]

# Extract the season and teams from the row
season = max_diff_row['Season'].values[0]   # Replace 'Season' with the actual column name in your DataFrame
winning_team = max_diff_row['Wteam'].values[0]  # Replace 'Wteam' with the actual column name in your DataFrame
losing_team = max_diff_row['Lteam'].values[0]   # Replace 'Lteam' with the actual column name in your DataFrame

# Print the results
print("Season:", season)
print("Teams involved:", winning_team, "vs", losing_team)


Season: 1996
Teams involved: 1409 vs 1341


# Grouping

Another important function in Pandas is **groupby()**. This is a function that allows you to group entries by certain attributes (e.g Grouping entries by Wteam number) and then perform operations on them. The following function groups all the entries (games) with the same Wteam number and finds the mean for each group. 

In [33]:
df.groupby('Wteam')['Wscore'].mean().head()

Wteam
1101    78.111111
1102    69.893204
1103    75.839768
1104    75.825944
1105    74.960894
Name: Wscore, dtype: float64

This next command groups all the games with the same Wteam number and finds where how many times that specific team won at home, on the road, or at a neutral site

In [34]:
df.groupby('Wteam')['Wloc'].value_counts().head(9)

Wteam  Wloc
1101   H        12
       A         3
       N         3
1102   H       204
       A        73
       N        32
1103   H       324
       A       153
       N        41
Name: Wloc, dtype: int64

Each dataframe has a **values** attribute which is useful because it basically displays your dataframe in a numpy array style format

In [35]:
df.values

array([[1985, 20, 1228, ..., 'N', 0, 17],
       [1985, 25, 1106, ..., 'H', 0, 7],
       [1985, 25, 1112, ..., 'H', 0, 7],
       ...,
       [2016, 132, 1246, ..., 'N', 1, 5],
       [2016, 132, 1277, ..., 'N', 0, 4],
       [2016, 132, 1386, ..., 'N', 0, 13]], dtype=object)

Now, you can simply just access elements like you would in an array. 

In [36]:
df.values[0][0]

1985

**Q**: Group the DataFrame by season and find the average winning score for each season.

In [37]:
# Group the DataFrame by season and calculate the average winning score for each season
average_winning_score = df.groupby('Season')['Wscore'].mean()

# Print the average winning score for each season
print(average_winning_score)


Season
1985    74.723040
1986    74.813640
1987    77.993870
1988    79.773704
1989    81.728511
1990    80.846477
1991    82.733932
1992    79.992004
1993    79.508790
1994    81.160345
1995    80.096149
1996    78.205483
1997    76.300842
1998    77.229902
1999    76.043581
2000    76.310467
2001    77.088650
2002    76.963557
2003    75.795061
2004    74.971122
2005    74.693690
2006    74.587555
2007    74.829467
2008    74.962425
2009    74.033721
2010    74.652480
2011    74.374190
2012    73.379973
2013    72.840789
2014    76.101268
2015    72.680799
2016    78.295958
Name: Wscore, dtype: float64


**Q**: Group the DataFrame by winning team and find the maximum winning score for each team across all seasons.

In [38]:
# Group the DataFrame by winning team and calculate the maximum winning score for each team
max_winning_score = df.groupby('Wteam')['Wscore'].max()

# Print the maximum winning score for each team
print(max_winning_score)


Wteam
1101     95
1102    111
1103    109
1104    114
1105    114
       ... 
1460    136
1461    112
1462    125
1463    105
1464    115
Name: Wscore, Length: 364, dtype: int64


**Q**: Group the DataFrame by both season and winning team. Find the team with the highest average winning score for each season.

In [39]:
# Group the DataFrame by season and winning team, and calculate the average winning score for each combination
average_winning_score = df.groupby(['Season', 'Wteam'])['Wscore'].mean()

# Find the team with the highest average winning score for each season
max_avg_winning_score = average_winning_score.groupby('Season').idxmax()

# Get the team ID with the highest average winning score for each season
teams_with_highest_avg_score = max_avg_winning_score.apply(lambda x: x[1])

# Print the team with the highest average winning score for each season
print(teams_with_highest_avg_score)


Season
1985    1328
1986    1109
1987    1380
1988    1258
1989    1258
1990    1258
1991    1380
1992    1380
1993    1380
1994    1380
1995    1206
1996    1206
1997    1254
1998    1395
1999    1317
2000    1395
2001    1395
2002    1242
2003    1395
2004    1194
2005    1311
2006    1190
2007    1440
2008    1440
2009    1377
2010    1440
2011    1449
2012    1370
2013    1377
2014    1322
2015    1322
2016    1146
Name: Wscore, dtype: int64


**Q**: Create a new DataFrame that counts the number of wins for each team in each season. This will involve grouping by both season and winning team, and then using the count() function.

In [40]:
# Group the DataFrame by season and winning team, and count the number of wins for each combination
wins_count = df.groupby(['Season', 'Wteam']).count()

# Reset the index of the resulting DataFrame to make the grouped columns regular columns
wins_count = wins_count.reset_index()

# Rename the count column to 'Wins'
wins_count = wins_count.rename(columns={'Wscore': 'Wins'})

# Print the new DataFrame with the count of wins for each team in each season
print(wins_count)

       Season  Wteam  Daynum  Wins  Lteam  Lscore  Wloc  Numot  \
0        1985   1102       5     5      5       5     5      5   
1        1985   1103       9     9      9       9     9      9   
2        1985   1104      21    21     21      21    21     21   
3        1985   1106      10    10     10      10    10     10   
4        1985   1108      19    19     19      19    19     19   
...       ...    ...     ...   ...    ...     ...   ...    ...   
10167    2016   1460      20    20     20      20    20     20   
10168    2016   1461      12    12     12      12    12     12   
10169    2016   1462      27    27     27      27    27     27   
10170    2016   1463      21    21     21      21    21     21   
10171    2016   1464       9     9      9       9     9      9   

       ScoreDifference  
0                    5  
1                    9  
2                   21  
3                   10  
4                   19  
...                ...  
10167               20  
10168  

**Q**: For each season, find the team with the most wins. This will involve creating a DataFrame similar to the one in task 5, and then using the idxmax() function for each season.

In [41]:
# Group the wins count DataFrame by 'Season'
grouped_by_season = wins_count.groupby('Season')

# Create an empty DataFrame to store the team with the most wins for each season
most_wins_per_season = pd.DataFrame(columns=['Season', 'Wteam', 'Wins'])

# Iterate over each season and find the team with the most wins using idxmax()
for season, group in grouped_by_season:
    max_wins_index = group['Wins'].idxmax()
    team_with_most_wins = group.loc[max_wins_index]
    most_wins_per_season = most_wins_per_season.append(team_with_most_wins)

# Reset the index of the resulting DataFrame
most_wins_per_season = most_wins_per_season.reset_index(drop=True)

# Print the DataFrame with the team with the most wins for each season
print(most_wins_per_season)


   Season Wteam Wins  Daynum  Lteam  Lscore  Wloc  Numot  ScoreDifference
0    1985  1385   27    27.0   27.0    27.0  27.0   27.0             27.0
1    1986  1181   32    32.0   32.0    32.0  32.0   32.0             32.0
2    1987  1424   33    33.0   33.0    33.0  33.0   33.0             33.0
3    1988  1112   31    31.0   31.0    31.0  31.0   31.0             31.0
4    1989  1328   28    28.0   28.0    28.0  28.0   28.0             28.0
5    1990  1247   29    29.0   29.0    29.0  29.0   29.0             29.0
6    1991  1116   30    30.0   30.0    30.0  30.0   30.0             30.0
7    1992  1181   28    28.0   28.0    28.0  28.0   28.0             28.0
8    1993  1231   28    28.0   28.0    28.0  28.0   28.0             28.0
9    1994  1163   27    27.0   27.0    27.0  27.0   27.0             27.0
10   1995  1116   26    26.0   26.0    26.0  26.0   26.0             26.0
11   1996  1269   31    31.0   31.0    31.0  31.0   31.0             31.0
12   1997  1242   31    31.0   31.0   

/var/folders/jp/76_36xgd3h55bg0npgjn08100000gn/T/ipykernel_15948/865499481.py:11: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  most_wins_per_season = most_wins_per_season.append(team_with_most_wins)
/var/folders/jp/76_36xgd3h55bg0npgjn08100000gn/T/ipykernel_15948/865499481.py:11: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  most_wins_per_season = most_wins_per_season.append(team_with_most_wins)
/var/folders/jp/76_36xgd3h55bg0npgjn08100000gn/T/ipykernel_15948/865499481.py:11: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  most_wins_per_season = most_wins_per_season.append(team_with_most_wins)
/var/folders/jp/76_36xgd3h55bg0npgjn08100000gn/T/ipykernel_15948/865499481.py:11: FutureWarning: The frame.append method is deprecate

**Q**: Group the DataFrame by losing team and find the average losing score for each team across all seasons. Compare this with the average winning score for each team from task 3. Are there teams that have a higher average losing score than winning score?

In [42]:
# Group the DataFrame by losing team and calculate the average losing score for each team
average_losing_score = df.groupby('Lteam')['Lscore'].mean()

# Print the average losing score for each team
print(average_losing_score)


Lteam
1101    60.586207
1102    59.201507
1103    64.117347
1104    64.374317
1105    61.675373
          ...    
1460    63.082915
1461    64.661972
1462    68.216117
1463    60.829213
1464    65.187063
Name: Lscore, Length: 364, dtype: float64


# Dataframe Iteration

In order to iterate through dataframes, we can use the **iterrows()** function. Below is an example of what the first two rows look like. Each row in iterrows is a Series object

In [44]:
for index, row in df.iterrows():
    print(row)
    if index == 1:
        break

Season             1985
Daynum               20
Wteam              1228
Wscore               81
Lteam              1328
Lscore               64
Wloc                  N
Numot                 0
ScoreDifference      17
Name: 0, dtype: object
Season             1985
Daynum               25
Wteam              1106
Wscore               77
Lteam              1354
Lscore               70
Wloc                  H
Numot                 0
ScoreDifference       7
Name: 1, dtype: object


**Q**: Create a new column 'HighScoringGame' that is 'Yes' if the winning score is greater than 100 and 'No' otherwise. This will require iterating over the rows of the DataFrame and checking the value of the winning score for each row.

In [45]:
# Iterate over the rows of the DataFrame
for index, row in df.iterrows():
    # Check if the winning score is greater than 100
    if row['Wscore'] > 100:
        df.at[index, 'HighScoringGame'] = 'Yes'  # Set 'Yes' in 'HighScoringGame' column
    else:
        df.at[index, 'HighScoringGame'] = 'No'   # Set 'No' in 'HighScoringGame' column

# Print the updated DataFrame
print(df)


        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
0         1985      20   1228      81   1328      64    N      0   
1         1985      25   1106      77   1354      70    H      0   
2         1985      25   1112      63   1223      56    H      0   
3         1985      25   1165      70   1432      54    H      0   
4         1985      25   1192      86   1447      74    H      0   
...        ...     ...    ...     ...    ...     ...  ...    ...   
145284    2016     132   1114      70   1419      50    N      0   
145285    2016     132   1163      72   1272      58    N      0   
145286    2016     132   1246      82   1401      77    N      1   
145287    2016     132   1277      66   1345      62    N      0   
145288    2016     132   1386      87   1433      74    N      0   

        ScoreDifference HighScoringGame  
0                    17              No  
1                     7              No  
2                     7              No  
3              

**Q**: Calculate the total number of games played by each team, whether they won or lost. This will require iterating over the rows of the DataFrame and updating a dictionary that keeps track of the number of games for each team.

In [46]:
# Create an empty dictionary to keep track of the number of games for each team
games_count = {}

# Iterate over the rows of the DataFrame
for index, row in df.iterrows():
    # Get the winning team and update the count
    winning_team = row['Wteam']
    if winning_team in games_count:
        games_count[winning_team] += 1
    else:
        games_count[winning_team] = 1

    # Get the losing team and update the count
    losing_team = row['Lteam']
    if losing_team in games_count:
        games_count[losing_team] += 1
    else:
        games_count[losing_team] = 1

# Print the total number of games played by each team
for team, count in games_count.items():
    print(f"Team {team} played {count} games.")

Team 1228 played 992 games.
Team 1328 played 968 games.
Team 1106 played 855 games.
Team 1354 played 906 games.
Team 1112 played 981 games.
Team 1223 played 363 games.
Team 1165 played 833 games.
Team 1432 played 69 games.
Team 1192 played 908 games.
Team 1447 played 903 games.
Team 1218 played 931 games.
Team 1337 played 922 games.
Team 1226 played 847 games.
Team 1242 played 993 games.
Team 1268 played 969 games.
Team 1260 played 914 games.
Team 1133 played 949 games.
Team 1305 played 922 games.
Team 1424 played 974 games.
Team 1307 played 969 games.
Team 1288 played 925 games.
Team 1344 played 951 games.
Team 1438 played 952 games.
Team 1374 played 916 games.
Team 1411 played 903 games.
Team 1412 played 962 games.
Team 1397 played 963 games.
Team 1417 played 966 games.
Team 1225 played 880 games.
Team 1116 played 980 games.
Team 1368 played 808 games.
Team 1120 played 936 games.
Team 1391 played 879 games.
Team 1135 played 847 games.
Team 1306 played 898 games.
Team 1143 played 947 

**Q**: For each season, find the game with the highest score difference (winning score - losing score). This will require iterating over the rows of the DataFrame, keeping track of the highest score difference for each season, and updating it if a game with a higher score difference is found.

In [47]:
# Create an empty dictionary to store the highest score difference for each season
highest_score_diff = {}

# Iterate over the rows of the DataFrame
for index, row in df.iterrows():
    season = row['Season']
    score_diff = row['Wscore'] - row['Lscore']

    if season in highest_score_diff:
        # Check if the current score difference is higher than the stored highest score difference for the season
        if score_diff > highest_score_diff[season]['score_diff']:
            highest_score_diff[season] = {'game': index, 'score_diff': score_diff}
    else:
        highest_score_diff[season] = {'game': index, 'score_diff': score_diff}

# Print the game with the highest score difference for each season
for season, data in highest_score_diff.items():
    game_index = data['game']
    score_diff = data['score_diff']
    print(f"In season {season}, the game with the highest score difference is game #{game_index} with a score difference of {score_diff}.")

In season 1985, the game with the highest score difference is game #236 with a score difference of 60.
In season 1986, the game with the highest score difference is game #4731 with a score difference of 84.
In season 1987, the game with the highest score difference is game #8240 with a score difference of 73.
In season 1988, the game with the highest score difference is game #12046 with a score difference of 68.
In season 1989, the game with the highest score difference is game #16677 with a score difference of 70.
In season 1990, the game with the highest score difference is game #19502 with a score difference of 76.
In season 1991, the game with the highest score difference is game #25161 with a score difference of 68.
In season 1992, the game with the highest score difference is game #27997 with a score difference of 82.
In season 1993, the game with the highest score difference is game #33858 with a score difference of 81.
In season 1994, the game with the highest score difference 

Remember, iterating over a DataFrame should generally be avoided if a vectorized operation can be used instead, as vectorized operations are usually much faster. However, these tasks are designed to give practice with DataFrame iteration for cases where it might be necessary.

Vectorized Operation Example: Create a new column 'HighScoringGame' in the DataFrame using a vectorized operation. This column should contain 'Yes' if the winning score is greater than 100 and 'No' otherwise. Use the np.where function from the numpy library for this task.

In [48]:
import numpy as np
df['HighScoringGame'] = np.where(df['Wscore'] > 100, 'Yes', 'No')

**Q**: Vectorized Operation: Calculate the total number of games played by each team, whether they won or lost. Instead of iterating over the DataFrame, use the value_counts() function on the winning team and losing team columns separately, and then add the two Series together.

In [49]:
# Calculate the total number of games played by each team using vectorized operations
games_count = df['Wteam'].value_counts() + df['Lteam'].value_counts()

# Print the total number of games played by each team
print(games_count)


1101     76
1102    840
1103    910
1104    975
1105    447
       ... 
1460    827
1461    914
1462    954
1463    838
1464    856
Length: 364, dtype: int64


**Q**: For each season, find the game with the highest score difference (winning score - losing score). Instead of iterating over the DataFrame, create a new column 'ScoreDifference' using vectorized subtraction, then use the groupby() function and idxmax() function to find the game with the highest score difference for each season.

In [50]:
# Create a new column 'ScoreDifference' using vectorized subtraction
df['ScoreDifference'] = df['Wscore'] - df['Lscore']

# Find the game with the highest score difference for each season
highest_score_diff_games = df.groupby('Season')['ScoreDifference'].idxmax()

# Print the game with the highest score difference for each season
for season, game_index in highest_score_diff_games.iteritems():
    score_diff = df.at[game_index, 'ScoreDifference']
    print(f"In season {season}, the game with the highest score difference is game #{game_index} with a score difference of {score_diff}.")
# Write your code here

In season 1985, the game with the highest score difference is game #236 with a score difference of 60.
In season 1986, the game with the highest score difference is game #4731 with a score difference of 84.
In season 1987, the game with the highest score difference is game #8240 with a score difference of 73.
In season 1988, the game with the highest score difference is game #12046 with a score difference of 68.
In season 1989, the game with the highest score difference is game #16677 with a score difference of 70.
In season 1990, the game with the highest score difference is game #19502 with a score difference of 76.
In season 1991, the game with the highest score difference is game #25161 with a score difference of 68.
In season 1992, the game with the highest score difference is game #27997 with a score difference of 82.
In season 1993, the game with the highest score difference is game #33858 with a score difference of 81.
In season 1994, the game with the highest score difference 

/var/folders/jp/76_36xgd3h55bg0npgjn08100000gn/T/ipykernel_15948/3919501788.py:8: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  for season, game_index in highest_score_diff_games.iteritems():


# Extracting Rows and Columns

The bracket indexing operator is one way to extract certain columns from a dataframe.

In [38]:
df[['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Notice that you can acheive the same result by using the loc function. Loc is a veryyyy versatile function that can help you in a lot of accessing and extracting tasks. 

In [39]:
df.loc[:, ['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Note the difference is the return types when you use brackets and when you use double brackets. 

In [40]:
type(df['Wscore'])

pandas.core.series.Series

In [41]:
type(df[['Wscore']])

pandas.core.frame.DataFrame

You've seen before that you can access columns through df['col name']. You can access rows by using slicing operations. 

In [42]:
df[0:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

Here's an equivalent using iloc

In [43]:
df.iloc[0:3,:]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

# Data Cleaning

One of the big jobs of doing well in Kaggle competitions is that of data cleaning. A lot of times, the CSV file you're given (especially like in the Titanic dataset), you'll have a lot of missing values in the dataset, which you have to identify. The following **isnull** function will figure out if there are any missing values in the dataframe, and will then sum up the total for each column. In this case, we have a pretty clean dataset.

In [44]:
df.isnull().sum()

Season    0
Daynum    0
Wteam     0
Wscore    0
Lteam     0
Lscore    0
Wloc      0
Numot     0
dtype: int64

If you do end up having missing values in your datasets, be sure to get familiar with these two functions. 
* **dropna()** - This function allows you to drop all(or some) of the rows that have missing values. 
* **fillna()** - This function allows you replace the rows that have missing values with the value that you pass in.

# Other Useful Functions

* **drop()** - This function removes the column or row that you pass in (You also have the specify the axis). 
* **agg()** - The aggregate function lets you compute summary statistics about each group
* **apply()** - Lets you apply a specific function to any/all elements in a Dataframe or Series
* **get_dummies()** - Helpful for turning categorical data into one hot vectors.
* **drop_duplicates()** - Lets you remove identical rows

# Lots of Other Great Resources

Pandas has been around for a while and there are a lot of other good resources if you're still interested on getting the most out of this library. 
* http://pandas.pydata.org/pandas-docs/stable/10min.html
* https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python
* http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/
* https://www.dataquest.io/blog/pandas-python-tutorial/
* https://drive.google.com/file/d/0ByIrJAE4KMTtTUtiVExiUGVkRkE/view
* https://www.youtube.com/playlist?list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y